From 74cbb0c09e70a6742579b340a4a0d4de3f5802f3 Mon Sep 17 00:00:00 2001
From: "Rywacki, Kamil" <kamil.rywacki@intel.com>
Date: Thu, 21 Jan 2021 12:48:15 +0100
Subject: [PATCH] PONRTSYS-8395: Fix Ethtool counters

- Remove re-inserted Broadcast, Unicast, Multicast and Good packets
- Remove re-inserted Bytes
- Fix re-inserted packets length calcuation
---
 drivers/net/datapath/dpm/datapath.h                 | 10 ++++++++++
 drivers/net/datapath/dpm/gswip31/datapath_tx.c      | 17 +++++++++++++++--
 drivers/net/datapath/mib_counters/dp_mib_counters.c |  5 +++++
 3 files changed, 30 insertions(+), 2 deletions(-)

diff --git a/drivers/net/datapath/dpm/datapath.h b/drivers/net/datapath/dpm/datapath.h
index 527e7c6045bc..150310fa0546 100644
--- a/drivers/net/datapath/dpm/datapath.h
+++ b/drivers/net/datapath/dpm/datapath.h
@@ -424,6 +424,16 @@ struct dp_reinsert_count {
 	u64 dp_MaxBytePkts;
 	/* DP Oversize Good Packet Count. */
 	u64 dp_OversizeGoodPkts;
+	/* DP Unicast Packet Count. */
+	u64 dp_UnicastPkts;
+	/* DP Multicast Packet Count. */
+	u64 dp_MulticastPkts;
+	/* DP Broadcast Packet Count. */
+	u64 dp_BroadcastPkts;
+	/* DP Good Packet Count. */
+	u64 dp_GoodPkts;
+	/* DP Good Bytes Count. */
+	u64 dp_GoodBytes;
 };
 
 struct logic_dev {
diff --git a/drivers/net/datapath/dpm/gswip31/datapath_tx.c b/drivers/net/datapath/dpm/gswip31/datapath_tx.c
index ca61730942ee..d92b906bd0ab 100644
--- a/drivers/net/datapath/dpm/gswip31/datapath_tx.c
+++ b/drivers/net/datapath/dpm/gswip31/datapath_tx.c
@@ -18,7 +18,7 @@ static int dp_reinsert_pkt_cnt(struct pmac_port_info *dp_info, int vap,
 	struct dp_reinsert_count *stats;
 	int inst = 0;
 	struct hal_priv *priv = HAL(inst);
-	int len = (skb->len - PMAC_SIZE);
+	int len = skb->len + ETH_FCS_LEN;
 
 	DP_DEBUG(DP_DBG_FLAG_DBG, "len=%d vap=%d\n", len, vap);
 	stats = &get_dp_port_subif(dp_info, vap)->reins_cnt;
@@ -38,6 +38,17 @@ static int dp_reinsert_pkt_cnt(struct pmac_port_info *dp_info, int vap,
 	} else if (len >= 1519) {
 		stats->dp_OversizeGoodPkts++;
 	}
+
+	if (is_broadcast_ether_addr(skb->data))
+		stats->dp_BroadcastPkts++;
+	else if (is_multicast_ether_addr(skb->data))
+		stats->dp_MulticastPkts++;
+	else
+		stats->dp_UnicastPkts++;
+
+	stats->dp_GoodPkts++;
+	stats->dp_GoodBytes += len;
+
 	DP_TX_CB_UNLOCK(&priv->cb_lock);
 	return DP_SUCCESS;
 }
@@ -309,9 +320,11 @@ int32_t dp_xmit_31(struct net_device *rx_if, dp_subif_t *rx_subif,
 		data.dp_inst = inst;
 		data.dp_inst = 0;
 	}
-	res = cbm_cpu_pkt_tx(skb, &data, 0);
+
 	if (flags & DP_TX_INSERT)
 		dp_reinsert_pkt_cnt(dp_info, vap, skb);
+
+	res = cbm_cpu_pkt_tx(skb, &data, 0);
 	UP_STATS(mib->tx_cbm_pkt);
 	MIB_G_STATS_INC(tx_pkts);
 	return res;
diff --git a/drivers/net/datapath/mib_counters/dp_mib_counters.c b/drivers/net/datapath/mib_counters/dp_mib_counters.c
index 5ba552e6b42e..3fe3f4fe3943 100644
--- a/drivers/net/datapath/mib_counters/dp_mib_counters.c
+++ b/drivers/net/datapath/mib_counters/dp_mib_counters.c
@@ -743,6 +743,11 @@ void mib_cnt_get_ethtool_stats_31(struct net_device *dev,
 	wide_cnt_mod.nRx1023BytePkts -= dp_reins_count.dp_1023BytePkts;
 	wide_cnt_mod.nRxMaxBytePkts -= dp_reins_count.dp_MaxBytePkts;
 	wide_cnt_mod.nRxOversizeGoodPkts -= dp_reins_count.dp_OversizeGoodPkts;
+	wide_cnt_mod.nRxUnicastPkts -= dp_reins_count.dp_UnicastPkts;
+	wide_cnt_mod.nRxBroadcastPkts -= dp_reins_count.dp_BroadcastPkts;
+	wide_cnt_mod.nRxMulticastPkts -= dp_reins_count.dp_MulticastPkts;
+	wide_cnt_mod.nRxGoodPkts -= dp_reins_count.dp_GoodPkts;
+	wide_cnt_mod.nRxGoodBytes -= dp_reins_count.dp_GoodBytes;
 
 	/* Copy counters values to ethtool buffer */
 	data_dst = data;
